File: 2016/day01/typescript/solution.ts
const directions = ["N", "E", "S", "W"] as const;

export function partOne(input: string) {
    const steps = parseInput(input);
    let x = 0;
    let y = 0;

    let direction: (typeof directions)[number] = directions[0];

    for (let step of steps) {
        let turn = step[0];
        let distance = parseInt(step.substring(1));

        if (turn === "R") {
            direction = directions[directions.indexOf(direction) + 1] || "N";
        } else {
            direction = directions[directions.indexOf(direction) - 1] || "W";
        }

        switch (direction) {
            case "N":
                y += distance;
                break;
            case "E":
                x += distance;
                break;
            case "S":
                y -= distance;
                break;
            case "W":
                x -= distance;
                break;
        }
    }

    return Math.abs(x) + Math.abs(y);
}

export function partTwo(input: string) {
    const steps = parseInput(input);
    let x = 0;
    let y = 0;

    let set = new Set<string>();

    let direction: (typeof directions)[number] = directions[0];

    for (let step of steps) {
        let turn = step[0];
        let distance = parseInt(step.substring(1));

        if (turn === "R") {
            direction = directions[directions.indexOf(direction) + 1] || "N";
        } else {
            direction = directions[directions.indexOf(direction) - 1] || "W";
        }

        for (let i = 0; i < distance; i++) {
            switch (direction) {
                case "N":
                    y++;
                    break;
                case "E":
                    x++;
                    break;
                case "S":
                    y--;
                    break;
                case "W":
                    x--;
                    break;
            }

            if (set.has(`${x},${y}`)) {
                return Math.abs(x) + Math.abs(y);
            }

            set.add(`${x},${y}`);
        }
    }

    throw new Error("never crossed paths");
}

function parseInput(input: string) {
    return input.split(/,\s/g);
}

File: 2016/day01/python/solution.py
directions = ["N", "E", "S", "W"]


def part_one(input):
    facing = "N"
    position = {'x': 0, 'y': 0}

    for step in input.strip().split(", "):
        turn = step[0]
        count = int(step[1:])

        if turn == "L":
            facing = directions[(directions.index(facing) - 1) %
                                4]
        elif turn == "R":
            facing = directions[(directions.index(facing) + 1) %
                                4]
        else:
            raise Exception("Invalid Turn")

        move(position, facing, count)

    return abs(position['x']) + abs(position['y'])


def part_two(input):
    facing = "N"
    position = {'x': 0, 'y': 0}
    visited = set()

    for step in input.strip().split(", "):
        turn = step[0]
        count = int(step[1:])

        if turn == "L":
            facing = directions[(directions.index(facing) - 1) %
                                4]
        elif turn == "R":
            facing = directions[(directions.index(facing) + 1) %
                                4]
        else:
            raise Exception("Invalid Turn")

        for _ in range(0, count):
            move(position, facing, 1)
            loc = f"{position['x']},{position['y']}"

            if loc in visited:
                return abs(position['x']) + abs(position['y'])
            else:
                visited.add(loc)


def move(position, direction, distance):
    if direction == "N":
        position['y'] += distance
    if direction == "E":
        position['x'] += distance
    if direction == "S":
        position['y'] -= distance
    if direction == "W":
        position['x'] -= distance


with open("../input", "r") as file:
    input = file.read()
    print("part one:", part_one(input))
    print("part two:", part_two(input))
